Malf AI powers are cooler now

Gilgaxx opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

Went ahead and fixed some things wrong with the Malf AI powers that irked me (that aren't the UI, that's a bit of a bigger task).

* Reactivate cameras Module: Removed one time purchase limitation, lowered the ammount of cameras fixed from 30 to 20, fixed a logging runtime (when using the last charge of the camera module), fixed the action description not updating the number of uses.
* This (previously non-working) feature of showing the uses left on a an action power was also given to the following powers (the other reusable ones): Overload Machine, Machine Override, Blackout and Fake Alert.
* For all 5 reusable actions I also made the unlock text and sound play when purchasing extra uses (previously had no feedback).

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

Fixing cameras isn't an extremly powerful tool for it to be unique (note that it doesn't rebuild them, it only fixes cameras that are offline that could be fixed with wirecutters). Plus most of the time if the crew is already cutting all of your cameras, repairing them should be the least of your worries compared to the imminent raid on your sattelite. For balance reasons, so if the crew keeps doing it to keep you from doing other things, I lowered the ammount of cameras fixed per processing power (30 -> 20 cameras for 10 processing power).
Feedback is also nice when purchasing things since the UI doesn't even update automatically to show that you've had processing power removed, so it's now nice to have a way to immedietly check that your purchase went through when buying more uses of a power you already had unlocked.
Finally fixing runtimes and broken features is nice.

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding.
-->

## Testing Photographs and Procedure
<!-- Include any screenshots/videos/debugging steps of the modified code functioning successfully, ideally including edge cases. -->
<details>
<summary>Screenshots (there's a few)</summary>
Started up my server as AI, spawned in a couple dozen broken cameras and made myself Malfunctioning.

The button to buy more of the cameras power is still there after a single purchase (no longer one time use):

![image](https://github.com/BeeStation/BeeStation-Hornet/assets/105242196/baae8db7-2952-42e2-b355-8bffacb69805)

No runtime when using all charges (the runtime on the screenshot was from me breaking a camera while setting up the screenshots):

![image](https://github.com/BeeStation/BeeStation-Hornet/assets/105242196/9e533ddb-d250-479e-a79d-36f5cf88c010)

Purchasing the same module twice in a row played the purchase sound effect and showed the text twice (to indicate the purchase was sucesfull twice), and hovering over the action button shows the updated number of lists correctly:

![image](https://github.com/BeeStation/BeeStation-Hornet/assets/105242196/13a2770c-18e7-497d-87e1-29ee820691db)

Another screenshot of the camera module after a partial use, showing the text in chat and the action button description correctly updating:

![image](https://github.com/BeeStation/BeeStation-Hornet/assets/105242196/60fb643f-7e5f-4467-9fe9-fac6aa5cee89)

Finally gave myself a few malf_upgrade disks to be able to purchase 2 of all the changed upgrades and saw that the action description was working for all of them

![image](https://github.com/BeeStation/BeeStation-Hornet/assets/105242196/b70f6d04-c5a0-439c-902a-ffaf21591bc9)

Also kept giving myself disks to be able to afford and use all other powers and used them just to double check everything else was unnafected (but I forgot to take a screenshot of this part, oopsie).

</details>

## Changelog
:cl:
balance: Removed one time limitation on Reactivate Cameras module for Malf AI but lowered the cameras fixed (30 -> 20)
fix: Action buttons for malf powers with multiple uses now correctly update
tweak: Purchasing more uses of a Malf Module now gives feedback in chat and with noise
fix: Removed a malf power logging runtime
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Gilgaxx committed:
Update to Malf Powers

Reactivate cameras: Fixed a logging runtime, fixed the action description not updating the number of uses, removed the one time purchase limitation, lowered the amount of uses.
Also added a counter to uses on the action description of the following powers(all the reusable ones): Overload Machine, Machine Override, Blackout and Fake Alert.
For these reusable actions I also made the unlock text and sound play when purchasing extra uses (previously had no feedback).
WARNING: UPDATING DESCRIPTION ON ACTIONS NOT WORKING ON PURCHASE, ONLY ON USE

Gilgaxx committed:
Fixes usage counter on two of the malf powers

I honestly don't know why I had to just add a -1 to the description counter but hey, it works.
WARNING: UPDATING DESCRIPTION ON ACTIONS NOT WORKING ON PURCHASE, ONLY ON USE

Gilgaxx committed:
Fixed the updating button description uses

Thank you Dakae :)

Gilgaxx committed:
Fixed the updating button description uses (again, oops)

I'm silly and forgot to change BOTH action.button.desc to action.desc. Should be good now.

PowerfulBacon commented on code/modules/antagonists/traitor/equipment/Malf_Modules.dm:
```diff
@@ -573,6 +575,11 @@ GLOBAL_LIST_INIT(blacklisted_malf_machines, typecacheof(list(
+/datum/action/innate/ai/ranged/overload_machine/New()
+	..()
+	desc = "[desc] There are [uses] reactivations remaining."
+	button.desc = desc
```
This should be handled in the parent type, inside of `/update_desc` by appending the number of uses to the examine message if the initial(uses) is different from 1. Then when you update the uses count, call update_appearance(UPDATE_DESC) instead

Gilgaxx replied:
Alright so, I made the requested changes, I think. I'm not super proficient yet, but I added a description of my reasoning to the commit, so let me know if I did anything wrong!

Gilgaxx committed:
Bacon's requested changes - making the description update be a parent thing

I added this proc:
```
/datum/action/innate/ai/proc/update_desc()
		desc = "[initial(desc)] There are [uses] reactivations remaining."
		button.desc = desc```
And then added `update_desc()` to the two `adjust_uses(amt, silent)`, all of the `/new` procs and then just removed a bunch of what appears to be redundant old code that wasn't doing anything (and getting in the way).

I also didn't include the "if the initial(uses) is different from 1" part you mentioned since fake alerts only has one use, but you can buy multiple and then you'd want the description to show that. I tried to get an if statement to check for the `one_purchase` variable, but that was in the modules themselves and I couldn't figure out how to check for it, so I just gave up.

I tested all the powers, including the RCD desctruction with the `one_purchase = true` and it didn't throw any runtimes or anything so it should be fine?

Gilgaxx committed:
Made the `/new()` stuff happen on the parent action

Made the `/new()` stuff happen on the parent action, cleaning up a bit of the code. with `if(initial(uses) > 1) update_desc()` on it, it works fine for normal powers, it doesn't proc for things like the RCD deconstruction, and the edge case of fake alarm, it doesn't show it has 1 use remaining unless you buy multiple copies, removing redundancy unless the AI wishes to actually stock piles fake alarms.

PowerfulBacon commented on code/modules/antagonists/traitor/equipment/Malf_Modules.dm:
```diff
@@ -51,8 +51,18 @@ GLOBAL_LIST_INIT(blacklisted_malf_machines, typecacheof(list(
+
+/datum/action/innate/ai/New()
+	..()
+	if(initial(uses) > 1)
```
You want to check for this inside update_desc instead, so that it only shows the uses after it updates if it has uses set.

Gilgaxx replied:
I thought so too, but like I said in the commit description this would make it so if you buy 2 fake alarms (which has an initial use of 1), and then use it once, the description wouldn't update and still say you had 2.

PowerfulBacon replied:
If you look at the default value for uses in `/datum/action/innate/ai`, the vaule is null which means that you dont need to check if its greater than 1, just if it has been set to anything at all. Then this will work.

Gilgaxx replied:
That's a good catch, but it seems that all `one_purchase` powers without the `UPGRADE` variable or the `auto_use_uses` variable actually have it set so they have `uses = 1`. Removing this breaks most powers, including making the RCD power infinite

Gilgaxx replied:
Actually the doomsday device doesn't have a use count, so the button never goes away after you use it. It just has: 
```
if (active)
	return
```

PowerfulBacon replied:
I see, in that case then I guess it doesn't matter too much

PowerfulBacon commented on code/modules/antagonists/traitor/equipment/Malf_Modules.dm:
```diff
@@ -919,6 +928,7 @@ GLOBAL_LIST_INIT(blacklisted_malf_machines, typecacheof(list(
 	var/datum/round_event/event_announcement = new event_control.typepath()
 	event_announcement.kill()
 	event_announcement.announce(TRUE)
+	owner_AI.update_action_buttons()
```
Do you need to do this here?

PowerfulBacon commented on code/modules/antagonists/traitor/equipment/Malf_Modules.dm:
```diff
@@ -888,7 +897,7 @@ GLOBAL_LIST_INIT(blacklisted_malf_machines, typecacheof(list(
 
 
-//Fake Alert: Overloads a random number of lights across the station. Three uses.
+//Fake Alert: Overloads a random number of lights across the station. Three uses. //Actually one use, always has been, silly comment
```
```suggestion
//Fake Alert: Overloads a random number of lights across the station. One use.
```

PowerfulBacon commented on code/modules/antagonists/traitor/equipment/Malf_Modules.dm:
```diff
@@ -761,6 +774,7 @@ GLOBAL_LIST_INIT(blacklisted_malf_machines, typecacheof(list(
 		else
 			apc.overload++
 	owner.log_message("activated malf module [name]", LOG_GAME)
+	owner_AI.update_action_buttons()
```
Does this need to be called here?

PowerfulBacon commented on code/modules/antagonists/traitor/equipment/Malf_Modules.dm:
```diff
@@ -176,21 +187,23 @@ GLOBAL_LIST_INIT(blacklisted_malf_machines, typecacheof(list(
 						action.uses += initial(action.uses)
 						temp = "Additional use[action.uses > 1 ? "s" : ""] added to [action.name]!"
+						action.desc = "[initial(action.desc)] There are [action.uses] reactivations remaining."
+						A.update_action_buttons()
```
```suggestion
						action.update_desc()
```

PowerfulBacon reviewed (changes requested):
I didn't realise update_desc was part of /atom, but this is an ideal implementation.

You are updating all of the user's action buttons in random places, although I dont believe that you need to. It should be called when the player loses access to their button, the description of the action and the button are updated when you change the uses appropriately. In regards to changing the description, all that you need to do is update the desc of the button, which is handled by UpdateButtonIcon when you call user.update_buttons so you don't need to do it there either.

Gilgaxx committed:
Bacon's requests 2

Changed the silly comment.

Replaced an old instances of manually changing an action description with `action.update_desc()`

removed 3 instances of redundant `action.update_desc()`

Gilgaxx committed:
Added a chat message when you can't afford modules

I noticed that it wasn't super obvious when you couldn't afford a module other than it changing the text at the bottom of the window (that is actually hidden in my UI)

Side note: this entire thing just needs a new UI, but that is outside the scope of this PR.

PowerfulBacon reviewed (commented):
Looks fine apart from the review that I added a comment to to clarify

Gilgaxx committed:
Fixes plural grammar

changes `"[initial(desc)] There are [uses] reactivations remaining."` to `"[initial(desc)] There [uses > 1 ? "are" : "is"] <b>[uses]</b> reactivation[uses > 1 ? "s" : ""] remaining."`, so it uses singular grammar when there's just one use left.
